feat: sales invoice integration with pos

diptanilsaha opened the pull request:
### Motivation

ERPNext POS uses POS Invoice and all the G/L Entries and Stock Ledger Entries are posted at the time of POS Closing Entry generation. However, several users requested for real-time update of G/L Entries and Stock Ledger Entries. Integrating the Sales Invoice in POS seems to be the best option to enable the real-time update of G/L Entries and Stock Ledger Entries.

### Changes Made

- Added a check field in Accounts Settings DocType on POS tab named `use_sales_invoice_in_pos`. By default, it is disabled. On enabling it, POS will load Sales Invoice instead of POS Invoice. Also, sales POS Invoice will not be allowed to be created however, return of POS Invoices will be allowed.
![image](https://github.com/user-attachments/assets/f290a5c6-78bb-4e5c-bbd7-6572a297a996)
- Added two hidden fields on Sales Invoice DocType named `is_created_using_pos` (check field) and `pos_closing_entry`(link field).  If a Sales Invoice is created using POS, `is_created_using_pos` will be enabled. At the time of the Consolidation of Sales Invoices, the `pos_closing_entry` will be filled.
- Added separate Child Table for Sales Invoice Transactions on POS Closing Entry DocType. If `use_sales_invoice_in_pos` is enabled, the POS Transactions child table will be hidden.
![image](https://github.com/user-attachments/assets/997c036d-b693-4ce6-8090-0d343751d652)
- A situation can arise where after enabling the Sales Invoice mode, a customer wants to return their purchase which was made using POS Invoice. In that case, the return will be made in POS Invoice and that POS Invoice will be consolidated as soon as the `Complete Order` button is clicked on POS. The consolidated Sales Invoice will be added in the POS Closing Entry.

### How to use Sales Invoice in POS

1. Make sure all the POS Opening Entries are closed.
2. Head over to `Accounts Settings` > `POS` tab and enable the `Use Sales Invoice`.
3. Now head over to Point of Sale and create a new POS Opening Entry.
4. POS will automatically create transactions using Sales Invoice.

For more information, please stay updated on: https://docs.frappe.io/erpnext/user/manual/en/point-of-sales#4-upcoming-features

diptanilsaha committed:
feat: pos configuration to activate real time update of gl and stock ledger

diptanilsaha committed:
feat: sales invoice on pos order list

diptanilsaha committed:
fix: syntax

diptanilsaha committed:
feat: past order list with sales invoice

diptanilsaha committed:
feat: customer recent transaction with sales invoice

diptanilsaha committed:
fix: real_time_update validation

added a validation to restrict switching between sales invoice and pos invoice in case there's already a pos opening entry

diptanilsaha committed:
fix: use sales invoice on accounts settings

moved the check to use sales invoice instead of pos invoice from pos profile to accounts settings.

diptanilsaha committed:
fix: using accounts settings to get frm doctype

diptanilsaha committed:
fix: added support for sales invoice in  process return

diptanilsaha committed:
feat: event listeners for sales invoice on pos

diptanilsaha committed:
fix: create and load return invoice

diptanilsaha committed:
fix: edit order

diptanilsaha committed:
refactor: function rename

diptanilsaha committed:
fix: sales invoice generation using pos

added fields to distinguish sales invoice generated using pos

diptanilsaha committed:
feat: credit note in pos invoice during sales invoice mode

diptanilsaha committed:
feat: pos closing entry support for sales invoices

Pranavrch commented:
Is it for v15 or v16?

diptanilsaha commented:
@Pranavrch v16

diptanilsaha committed:
refactor: resolving linter issues

diptanilsaha committed:
refactor: fix linter issue

diptanilsaha committed:
fix: filters for sales invoice in toggle recent orders

diptanilsaha committed:
feat: disable partial payments on sales invoice transactions made using pos

Pranavrch commented:
Hey. What is gonna happen to the print format? We'll have to make one for sales invoice right?

diptanilsaha commented:
> Hey. What is gonna happen to the print format? We'll have to make one for sales invoice right?

You can create one and set it on the POS Profile.

diptanilsaha committed:
fix: resolve import error

diptanilsaha committed:
fix: recent order list and pos invoice returns during sales invoice mode

diptanilsaha committed:
fix: reset pos_closing_entry on return sales invoice

diptanilsaha committed:
fix: filtering out consolidated return sales invoice for pos invoice return

diptanilsaha committed:
fix: pos delete order

diptanilsaha committed:
fix: added missing reference to consolidated sales invoice item

diptanilsaha committed:
fix: added check to restrict sales invoice creation

dineshpanchal93 commented:
Waiting for the release !!!

diptanilsaha committed:
refactor: variable name

worldinet commented:
This is really great, looking forward to it! But I was wondering if we'll have to change our client script doctype to Sales Invoice in place of POS Invoice?

diptanilsaha committed:
fix: integrating sales_invoice in make_closing_entry_from_opening method

diptanilsaha committed:
test: test for sales invoice integration in pos

diptanilsaha commented:
> This is really great, looking forward to it! But I was wondering if we'll have to change our client script doctype to Sales Invoice in place of POS Invoice?

Can you explain what "client script" is about?

worldinet commented:
> > This is really great, looking forward to it! But I was wondering if we'll have to change our client script doctype to Sales Invoice in place of POS Invoice?
> 
> Can you explain what "client script" is about?

Sorry, I wasn't clear first. I meant if you have a client script running on existing POS Invoice doctype and you enable the use sales invoice option, will I have to modify the script for Sales Invoice doctype instead?

diptanilsaha commented:
> > > This is really great, looking forward to it! But I was wondering if we'll have to change our client script doctype to Sales Invoice in place of POS Invoice?
> > 
> > 
> > Can you explain what "client script" is about?
> 
> Sorry, I wasn't clear first. I meant if you have a client script running on existing POS Invoice doctype and you enable the use sales invoice option, will I have to modify the script for Sales Invoice doctype instead?

I think that you have to figure it out on your own because it totally depends on your use case. Also, once you enable "Use Sales Invoice" feature, you won't be able to create a sales POS Invoice.

kittiu commented:
@diptanilsaha thanks for pointing this. But if I am not mistaken, the reason to consolidated pos invoices is for performance improvement (as the previous versions of ERPNext having this very problem before the merge log).

Will this creation of Sales Invoice and do the posting immediatelly casuing performance issues for high volume transactions?

If the performance is not affected, for me, this is a straight forward ways to work with, and should remove some issues we found with POS Invoice.

diptanilsaha commented:
@kittiu You are correct regarding the Performance Issue and this feature is only meant for low-volume transactions. Also, please report the issues you found on the POS Invoice.

muchai commented:
@diptanilsaha Use Sales Invoice is a toggle, right? Say I chose to use Sales Invoices today, then after a while I want to revert to use POS Invoices for performance, this is possible, right?

diptanilsaha commented:
> @diptanilsaha Use Sales Invoice is a toggle, right? Say I chose to use Sales Invoices today, then after a while I want to revert to use POS Invoices for performance, this is possible, right?

Yes, 'Use Sales Invoice' is a toggle and you can go back to using POS Invoice. I will update the documentation soon.

nabinhait commented on erpnext/accounts/doctype/accounts_settings/accounts_settings.json:
```diff
@@ -532,14 +534,26 @@
+   "description": "If enabled, Sales Invoice will be generated instead of POS Invoice in POS Transactions for real-time update of G/L and Stock Ledger.",
+   "fieldname": "use_sales_invoice_in_pos",
+   "fieldtype": "Check",
+   "label": "Use Sales Invoice"
```
```suggestion
   "label": "Create Sales Invoice directly from the POS"
```

nabinhait commented on erpnext/accounts/doctype/pos_closing_entry/pos_closing_entry.js:
```diff
@@ -113,7 +123,25 @@ frappe.ui.form.on("POS Closing Entry", {
+		});
+	},
+
+	get_sales_invoices(frm) {
```
`get_pos_invoices` and `get_sales_invoice` function should be commonified

nabinhait commented on erpnext/accounts/doctype/pos_closing_entry/pos_closing_entry.js:
```diff
@@ -132,18 +160,49 @@ frappe.ui.form.on("POS Closing Entry", {
+		if (is_pos_using_sales_invoice) {
+			await Promise.all([
+				frappe.call({
+					method: "erpnext.accounts.doctype.pos_closing_entry.pos_closing_entry.get_pos_invoices",
```
This is repetitive, you can reuse get_pos_invoices function and then calculate the totals after that.

THE BETTER:
Move everything to serverside. On saving the document, we can fetch pos invoices or sales invoices, calculate the totals etc inside the validate method. Even fetching those on the selection of pos_opening_entry is not very useful, as we are not giving any option to modify the invoice list.

nabinhait commented on erpnext/accounts/doctype/pos_closing_entry/pos_closing_entry.json:
```diff
@@ -227,16 +228,23 @@
    "reqd": 1
+  },
+  {
+   "fieldname": "sales_invoice_transactions",
```
```suggestion
   "fieldname": "sales_invoices",
```

nabinhait commented on erpnext/accounts/doctype/pos_closing_entry/pos_closing_entry.py:
```diff
@@ -58,8 +60,20 @@ def validate(self):
 
-		self.validate_duplicate_pos_invoices()
-		self.validate_pos_invoices()
+		self.is_pos_using_sales_invoice = frappe.db.get_single_value(
+			"Accounts Settings", "use_sales_invoice_in_pos"
+		)
+
+		if self.is_pos_using_sales_invoice == 0:
+			self.validate_duplicate_pos_invoices()
+			self.validate_pos_invoices()
+
+		if self.is_pos_using_sales_invoice == 1:
+			if len(self.pos_transactions) != 0:
+				frappe.throw(_("POS Invoices can't be added when Sales Invoice is enabled"))
+
+		self.validate_duplicate_sales_invoices()
+		self.validate_sales_invoices()
```
Move it to a wrapper function (no direct code inside the validate, only function calls).

Even setting posting_date and time to now and validating the opening entry should be separate functions.

nabinhait commented on erpnext/accounts/doctype/pos_closing_entry/pos_closing_entry.py:
```diff
@@ -143,6 +226,12 @@ def update_opening_entry(self, for_cancel=False):
 
+	def update_sales_invoices_closing_entry(self, cancel=False):
+		for d in self.sales_invoice_transactions:
+			frappe.db.set_value(
```
Can we bulk update?

nabinhait commented on erpnext/accounts/doctype/pos_profile/pos_profile.json:
```diff
@@ -407,16 +407,9 @@
    "fieldtype": "Check",
    "label": "Disable auto setting Grand Total to default Payment Mode"
-  },
-  {
```
Why removed?

kittiu commented:
> this

It is ok, I think we can live with this in trade of simplicity. For the on demand full sales invoice, I think we can make custom module case by case. Thanks!

diptanilsaha committed:
fix: issue with accounting dimension on sales invoice

diptanilsaha committed:
refactor: moved invoice switching mode validation in backend

diptanilsaha committed:
test: removed test case

Removed Test Case for Full Payment of Sales Invoice created using POS as planning to add feature to accept Partial Payment from POS.

diptanilsaha committed:
test: fixed the failing tests

diptanilsaha committed:
test: remove explicit use of frappe.db.commit()

diptanilsaha committed:
test: fixing pos invoice test

diptanilsaha committed:
test: removed test

mahsem commented:
> Is it for v15 or v16?

Could it backported to v15 ?

diptanilsaha commented:
> Could it backported to v15 ?

We are planning to release this feature in v16.

Pranavrch commented:
I tried it in the develop branch. If you're using other company for pos which isn't the default company, the payment account (in this case cash) is set to that of the default company's

Screenshot -
![Screenshot_2025-04-30-09-42-20-298_com android chrome](https://github.com/user-attachments/assets/d527597b-eef2-41ab-8cbc-abc8ad2b7e69)

diptanilsaha commented:
@Pranavrch Thank you for the feedback. I will look into it. In the future, please create a GitHub issue and tag me if it relates to POS.

Pranavrch commented:
Yes. I didn't know if we could create GitHub issues for develop branch. I'll use that in the future, and thank you for bringing these improvements to the pos your work is very helpful for the enhancement of erpnext retail experience!

diptanilsaha commented:
@Pranavrch could you please provide the steps to replicate the issue? It would be helpful to create a GitHub issue where you can share the steps and additional details.

Pranavrch commented:
Steps-

1. Create 2 or more companies.
2. Set company A as default in global defaults
3. Create a pos profile with company B
4. Open the pos and try creating an invoice 

You get this error -
Sales Invoice ACC-SINV-2025-05586: Account Cash - KU does not belong to Company Happy Moments
